linux/ibus-keyman/src/engine.c

```diff
@@ -595,25 +596,30 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
 
     g_message("DAR: ibus_keyman_engine_process_key_event - keyval=%02i, keycode=%02i, state=%02x", keyval, keycode, state);
 
+    gboolean isKeyDown = TRUE;
+
     if (state & IBUS_RELEASE_MASK)
     {
-        switch(keycode) {
-            case KEYMAN_LCTRL:
-                keyman->lctrl_pressed = FALSE;
-                break;
-            case KEYMAN_RCTRL:
-                keyman->rctrl_pressed = FALSE;
-                break;
-            case KEYMAN_LALT:
-                keyman->lalt_pressed = FALSE;
-                break;
-            case KEYMAN_RALT:
-                keyman->ralt_pressed = FALSE;
-                break;
-            default:
-                break;
+        switch (keycode)
+        {
+        case KEYMAN_LCTRL:
+            keyman->lctrl_pressed = FALSE;
+            return FALSE;
+        case KEYMAN_RCTRL:
+            keyman->rctrl_pressed = FALSE;
+            return FALSE;
+        case KEYMAN_LALT:
+            keyman->lalt_pressed = FALSE;
+            return FALSE;
+        case KEYMAN_RALT:
+            keyman->ralt_pressed = FALSE;
+            return FALSE;
+        case KEYMAN_CAPS:
+            isKeyDown = FALSE;
+            break;
+        default:
+            return FALSE;
         }
-        return FALSE;
     }
```

Comment:
We should be passing the key up events to the core now, for all keys, probably even modifier keys. While the kmx processor doesn't do much with them, we can't guarantee that's the case for future processors. Thus, something like this:

```c++
    if (state & IBUS_RELEASE_MASK) {
        isKeyDown = FALSE;
        switch(keycode) {
        case KEYMAN_LCTRL:
            keyman->lctrl_pressed = FALSE;
            break;
        case KEYMAN_RCTRL:
            keyman->rctrl_pressed = FALSE;
            break;
        case KEYMAN_LALT:
            keyman->lalt_pressed = FALSE;
            break;
        case KEYMAN_RALT:
            keyman->ralt_pressed = FALSE;
            break;
        }
    }

Reply:
done